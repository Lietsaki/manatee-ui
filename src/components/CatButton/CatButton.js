import React from 'react';
import { string, object, func } from 'prop-types';
import styled from 'styled-components';
import { getHSLA, getHSLADarker, getBeforeDimensions } from '@bit/lietsaki.manatee-ui.manatee-utilities';

const defaultColor = { hue: '270', saturation: '43', lightness: '52', alpha: '1' };
const defaultGradientColor1 = '#654ea3';
const defaultGradientColor2 = '#eaafc8';

// 1) Define the button body
const ButtonBody = styled.button`
  color: ${(props) => (props.textColor ? props.textColor : '#fff')};
  background-color: ${(props) => (props.bgColor ? getHSLA(props.bgColor) : getHSLA(defaultColor))};
  width: ${(props) => (props.width ? props.width : '120px')};
  height: ${(props) => (props.height ? props.height : '39px')};
  border-radius: 10rem;
  display: flex;
  justify-content: center;
  align-items: center;
  transition: all 0.3s ease-in-out;
  border: none;
  font-size: 15px;
  outline: none;
  position: relative;

  &:hover {
    cursor: pointer;
    background-color: ${(props) => (props.bgColor ? getHSLADarker(props.bgColor) : getHSLADarker(defaultColor))};

    svg {
      fill: ${(props) => (props.bgColor ? getHSLADarker(props.bgColor) : getHSLADarker(defaultColor))};
    }
  }

  &:active {
    transition: all 0.1s ease-in-out;
    transform: scale(0.97);
  }

  &:before {
    content: '';
    display: block;
    position: absolute;
    z-index: -20;
    width: ${(props) =>
      props.width ? getBeforeDimensions('width', props.width) : getBeforeDimensions('width', '120px')};
    height: ${(props) =>
      props.height ? getBeforeDimensions('height', props.height) : getBeforeDimensions('height', '37px')};
    border-radius: 2rem;
    transition: box-shadow 0.2s;
    will-change: box-shadow;
  }

  &:active:before,
  &:focus:before {
    box-shadow: -0.09375rem -0.09375rem 0 0.09375rem ${(props) => (props.beforeGradient1 ? props.beforeGradient1 : defaultGradientColor1)},
      0 0 0 0.1875rem ${(props) => (props.beforeGradient2 ? props.beforeGradient2 : defaultGradientColor2)};
  }
`;

// 2) Define the svg element with its attributes and styles
const Ear = styled.svg.attrs({
  version: '1.1',
  xmlns: 'http://www.w3.org/2000/svg',
  xmlnsXlink: 'http://www.w3.org/1999/xlink',
  viewBox: '0 0 1080 1080'
})`
  fill: ${(props) => (props.bgColor ? getHSLA(props.bgColor) : getHSLA(defaultColor))};
  width: 30px;
  height: 30px;
  position: absolute;
  z-index: -1;
  bottom: 55%;
  transition: all 0.3s ease-in-out;
`;

// 3) Define the path element - When we want to store some JSX in a variable, we do it in a function that returns such
// JSX, just like this. If we try do simply store the JSX block directly without using a function, an error will be thrown.
const EarPath = () => (
  <path
    d="M1053.4,868.4c51.1-98.9-38.9-242.6-156.5-430.3C779,249.7,660.6,60.7,505.8,66.5
c-141.5,5.2-240,170.4-371.6,391.2C43.6,609.5-1.7,685.4,16.7,751.1C89.3,1009.6,934.5,1098.7,1053.4,868.4z"
  />
);

// 4) Attach the path as a child of the SVG to create the full ear.
const FullEar = (props) => (
  <Ear style={props.style} bgColor={props.bgColor}>
    <EarPath />
  </Ear>
);

// 5) Example of styles used to position each ear (each svg element)
const rightEarStyles = {
  right: '15px',
  bottom: '55%'
};
const leftEarStyles = {
  left: '15px',
  bottom: '55%'
};

const getCatButton = (props) => {
  const {
    onClick,
    bgColor,
    textColor,
    width,
    height,
    beforeGradient1,
    beforeGradient2,
    rightEarStyles,
    leftEarStyles
  } = props;

  // Trigger event by pressing the enter key.
  const handleKeypress = (e) => {
    if (e.keyCode === 13) {
      onClick();
    }
  };

  return (
    <ButtonBody
      onClick={onClick}
      onKeyPress={handleKeypress}
      bgColor={bgColor}
      textColor={textColor}
      width={width}
      height={height}
      beforeGradient1={beforeGradient1}
      beforeGradient2={beforeGradient2}
    >
      <FullEar style={rightEarStyles} bgColor={bgColor} />
      <FullEar style={leftEarStyles} bgColor={bgColor} />
      {props.children}
    </ButtonBody>
  );
};

getCatButton.defaultProps = {
  rightEarStyles,
  leftEarStyles,
  bgColor: defaultColor,
  beforeGradient1: defaultGradientColor1,
  beforeGradient2: defaultGradientColor2,
  textColor: '#fff',
  width: '120px',
  height: '39px'
};

getCatButton.propTypes = {
  onClick: func,
  bgColor: object,
  textColor: string,
  width: string,
  height: string,
  beforeGradient1: string,
  beforeGradient2: string,
  leftEarStyles: object,
  rightEarStyles: object
};

export default getCatButton;
