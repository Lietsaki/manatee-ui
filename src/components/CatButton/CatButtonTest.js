import React from 'react';
import CatButton from './CatButton';

// To test all our customization options, let's create a button using every possible prop.
const garfieldOrange = { hue: '22', saturation: '84', lightness: '52', alpha: '1' };
const logMeow = () => console.log('Meow! You clicked the cat button.');

const CatButtonTest = () => {
  return (
    <>
      <CatButton>Log In</CatButton>
      <CatButton
        onClick={logMeow}
        bgColor={garfieldOrange}
        textColor="#151515"
        width="15vw"
        height="5vh"
        beforeGradient1="#FDC830"
        beforeGradient2="#F37335"
        rightEarStyles={{ right: '15px' }}
        leftEarStyles={{ left: '15px' }}
      >
        <p>Add to cart</p>
      </CatButton>
    </>
  );
};

export default CatButtonTest;
