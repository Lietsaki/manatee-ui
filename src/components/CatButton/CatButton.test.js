import React from 'react';
import renderer from 'react-test-renderer';
import CatButton from './CatButton';
import { render, fireEvent } from '@testing-library/react';

// 1) Define a mock callback for the button
const mockCallback = jest.fn();

// 2) Store the component in a variable
const button = (
  <CatButton onClick={mockCallback}>
    <p>Cancel</p>
  </CatButton>
);

// TEST #1: Rendering the button's content
test('Renders CatButton correctly', () => {
  // 1) Render the button with react-testing-library in order to be able to get it by its text.
  const { getByText } = render(button);
  getByText('Cancel');

  // 2) Create the element with renderer to create snapshots between each callback
  const rendererButton = renderer.create(button);

  // 3) Take a snapshot with Jest
  let tree = rendererButton.toJSON();
  expect(tree).toMatchSnapshot();
});

// TEST #2: onClick callback execution
test('Executes CatButton onClick callback', () => {
  // 1) Render the button with React Testing Library to trigger a click event
  const { getByRole } = render(button);
  const RTLButton = getByRole('button');
  fireEvent.click(RTLButton);

  // 2) Expect that the callback was called only once
  expect(mockCallback).toHaveBeenCalledTimes(1);
});

// NOTE: The tab + enter callback execution is not tested because of the limited support Jest currently has, but it's been tested
// in the browser and works perfectly as well.
