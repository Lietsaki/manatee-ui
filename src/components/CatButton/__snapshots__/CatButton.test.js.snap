// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders CatButton correctly 1`] = `
<button
  className="sc-bdfBwQ dRcJhu"
  height="39px"
  onClick={[MockFunction]}
  onKeyPress={[Function]}
  width="120px"
>
  <svg
    className="sc-gsTCUz cWpiuZ"
    style={
      Object {
        "bottom": "55%",
        "right": "15px",
      }
    }
    version="1.1"
    viewBox="0 0 1080 1080"
    xmlns="http://www.w3.org/2000/svg"
    xmlnsXlink="http://www.w3.org/1999/xlink"
  >
    <path
      d="M1053.4,868.4c51.1-98.9-38.9-242.6-156.5-430.3C779,249.7,660.6,60.7,505.8,66.5
c-141.5,5.2-240,170.4-371.6,391.2C43.6,609.5-1.7,685.4,16.7,751.1C89.3,1009.6,934.5,1098.7,1053.4,868.4z"
    />
  </svg>
  <svg
    className="sc-gsTCUz cWpiuZ"
    style={
      Object {
        "bottom": "55%",
        "left": "15px",
      }
    }
    version="1.1"
    viewBox="0 0 1080 1080"
    xmlns="http://www.w3.org/2000/svg"
    xmlnsXlink="http://www.w3.org/1999/xlink"
  >
    <path
      d="M1053.4,868.4c51.1-98.9-38.9-242.6-156.5-430.3C779,249.7,660.6,60.7,505.8,66.5
c-141.5,5.2-240,170.4-371.6,391.2C43.6,609.5-1.7,685.4,16.7,751.1C89.3,1009.6,934.5,1098.7,1053.4,868.4z"
    />
  </svg>
  <p>
    Cancel
  </p>
</button>
`;
